src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs

```diff
@@ -27,18 +27,33 @@ public static T[] GetSubArray<T>(T[] array, Range range)
 
             (int offset, int length) = range.GetOffsetAndLength(array.Length);
 
-            if (length == 0)
+            T[] dest;
+
+            if (typeof(T).IsValueType || typeof(T[]) == array.GetType())
             {
-                return Array.Empty<T>();
+                // We know the type of the array to be exactly T[] or an array variance
+                // compatible value type substitution like int[] <-> uint[].
+
+                if (length == 0)
+                {
+                    return Array.Empty<T>();
+                }
+
+                dest = new T[length];
             }
+            else
+            {
+                // The array is actually a U[] where U:T. We'll make sure to create
+                // an array of the exact same backing type. The cast to T[] will
+                // never fail.
 
-            T[] dest = new T[length];
+                dest = Unsafe.As<T[]>(Array.CreateInstanceFromArrayType(array.GetType(), length));
+            }
 
-            // Due to array variance, it's possible that the incoming array is
-            // actually of type U[], where U:T; or that an int[] <-> uint[] or
-            // similar cast has occurred. In any case, since it's always legal
-            // to reinterpret U as T in this scenario (but not necessarily the
-            // other way around), we can use Buffer.Memmove here.
+            // In either case, the newly-allocated array is the exact same type as the
```

Comment:
> is the exact same type

This isn't the case for the int[] <-> uint[] case, right?  That's not a problem, but maybe something to clarify in the comment?

Reply:
Good point. Does C# LDM want the exact type preserved even for compatible valuetype substitutions?

This prints `uint[]` today:
```
int[] a = new int[10];
uint[] b = (uint[])(object)a;
Console.WriteLine(b[1..2].GetType());
```
Should it print `int[]` instead?

In other words, should the actual type of array slice always match the type of the source array?

Reply:
Yes, I think this should print `int[]` to mimic a virtual call. Tagging @MadsTorgersen to confirm

Reply:
Ideally, I guess, but this seems like such a corner case that I doubt anyone would notice in practice, let alone be hurt by it. In the C# language case, since a runtime type check will succeed between `int[]` and `uint[]`, the motivating scenario will be back to working again either way.